charmcraft/commands/store/__init__.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""Commands related to the Store, a thin layer above real functionality."""
+
+import logging
+
+from charmcraft.cmdbase import BaseCommand
+from charmcraft.commands.store.client import Client
+
+logger = logging.getLogger('charmcraft.commands.store')
+
+
+class LoginCommand(BaseCommand):
+    """Log into the store."""
+    name = 'login'
+    help_msg = "login in the Ubuntu Single Sign On"
+
+    def run(self, parsed_args):
+        """Run the command.
+
+        The login happens on every request to the Store (if current credentials were not
+        enough), so here we just exercise the simplest command regarding developer identity.
+        """
+        client = Client()
+        client.clear_credentials()
+        client.get('/v1/whoami')
+        logger.info("Login successful")
+
+
+class LogoutCommand(BaseCommand):
+    """Clear store-related credentials."""
+    name = 'logout'
+    help_msg = "clear session credentials"
+
+    def run(self, parsed_args):
+        """Run the command."""
+        client = Client()
+        client.clear_credentials()
+        logger.info("Credentials cleared")
+
+
+class WhoamiCommand(BaseCommand):
+    """Show login information."""
+    name = 'whoami'
+    help_msg = "returns your login information relevant to the store"
+
+    _titles = [
+        ('name:', 'display-name'),
+        ('username:', 'username'),
+        ('id:', 'id'),
+    ]
+
+    def run(self, parsed_args):
+        """Run the command."""
+        client = Client()
+        result = client.get('/v1/whoami')
```

Comment:
Ultimately it feels like there should be a Python class that knows how to talk to the API and URLs without being the command line code.

iow this would be Client.whoami() which then does a URL /v1/whoami. It feels odd to have the run() methods encoding how to talk to the charmstore API.

Then the commonality between this and Login would be pulled out.

Generally python class talking to the API is very shareable and reusable code that should exist independent of the CLI code.

I'm ok to start here, but I feel like we either want Client to grow more, or Client is a raw client and we grow something above it.

Reply:
Yes, I'll add an `api` layer that knows how to talk to the Store, and remove from this particular file any "magic string". But I wanted to leave that to the next branch when adding more complex commands, as this branch introduced already too many moving parts.